### PR TITLE
scratchblocks: fix newlines and hopefully fix race condition

### LINKED ISSUE #3304: `scratchblocks` Addon Ignores Newline Characters
### Description

The `scratchblocks` addon does not work with scripts more than one block tall. It acts as though the newline characters doesn't exist.

### Repoduction steps

1. Enable the `scratchblocks` addon
2. Find a script that uses more than one block.
3. Notice how new line characters are ignored and the script is all on one line.

### Expected behaviour

Newline characters should not be ignored.

### Actual behaviour

Newline characters are ignored.

### Screenshots

Enabled:
Screenshot (19)

Disabled:
Screenshot (20)

### Environment

- Browser: Edge 92.0.902.73
- Scratch Addons version: 1.19.0 pre-realese
- Operating system: Windows 11 running build 22000.132

### Additional Info

Can repo with all addons on and with all addons off except for `scratchblocks`

### PR DESCRIPTION
**Pull request by FunctionalMetatable**
Resolves #3304

Fixes the newline character bug and race cond.

@apple502j

### CODE REVIEW

**Comment by apple502j** [comment]
#3306 is in no way related to #3304; the addon literally does not work for me. Sometimes the addon loads faster than cs.js, sometimes the addon loads but then original one loads and overwrites, sometimes the addon misses the blocks because they aren't loaded yet, etc.

**Comment by apple502j** [comment]
Doesn't fix the problem - it's even worse because sb3 is not loading at all. `data-original` is set, but original one is rendered. Also, `if (blocks.length !== 0)` looks wrong. And `blocks` seems to be empty on my side.

**Comment by apple502j** [comment]
Note that if I fixed both problems, it still doesn't work; this time it crashes, `svg is undefined @ userscript.js L2`

**Comment by FunctionalMetatable** [comment]
> I'd recommend testing the addon on Firefox, because it never works there.

It works for me (Firefox 91.0)

**Comment by apple502j** [comment]
@FunctionalMetatable Test on https://scratch.mit.edu/discuss/topic/514769/ ?

**Comment by FunctionalMetatable** [comment]
> @FunctionalMetatable Test on [scratch.mit.edu/discuss/topic/514769](https://scratch.mit.edu/discuss/topic/514769/) ?

thats where i was testing

**Comment by apple502j** [comment]
@FunctionalMetatable Still not working:

image

File: addons/scratchblocks/userscript.js [hunk 1]
```diff
@@ -44,16 +44,25 @@ async function getLocales(addon) {
...
-    block.textContent = block.getAttribute("data-original");
-  });
 
   // Translations can't load first
   await addon.tab.loadScript(addon.self.lib + "/thirdparty/cs/scratchblocks-v3.5.2-min.js");
   await addon.tab.loadScript(addon.self.lib + "/thirdparty/cs/translations-all-v3.5.2.js");
 
   const languages = await getLocales(addon);
 
+  if (blocks.length > 0) {
```
Comment (apple502j) [inline addons/scratchblocks/userscript.js:54]: I got `blocks.length === 0` at this line (remember: this addon can load before DOM!)

Comment (FunctionalMetatable) [reply addons/scratchblocks/userscript.js:62]: Now it gets the blocks after the DOM is loaded, try now?

**Comment by apple502j** [comment]
@FunctionalMetatable Crash not fixed

```js
Uncaught (in promise) TypeError: svg is undefined
    scaleSVG moz-extension://f3d7d0c6-7c48-4f57-851f-b8ffe33b98fb/addons/scratchblocks/userscript.js:2
    renderMatching moz-extension://f3d7d0c6-7c48-4f57-851f-b8ffe33b98fb/addons/scratchblocks/userscript.js:89
```

image
image

**Comment by apple502j** [comment]
@FunctionalMetatable This is another race condition; this causes crash because `scratchblocks.render` is v2 (not v3 that we loaded before), and the function returns different values in different versions.

**Comment by apple502j** [comment]
Error is now gone, but this time after v3 successfully renders, v2 gets loaded and rendered - sometimes (works 50%, fails 50%).


https://user-images.githubusercontent.com/33279053/129761290-71d0b96d-b44a-4533-8304-a0448aff8d9c.mp4

**Comment by apple502j** [comment]
@FunctionalMetatable ?

**Comment by FunctionalMetatable** [comment]
> v2 gets loaded and rendered

Should be fixed now, v2 now gets blocked by blocking writable

**Comment by apple502j** [comment]
@FunctionalMetatable Still inconsistent

image

image

**Comment by apple502j** [comment]
Still experiencing the same issue. Also
image

**Comment by FunctionalMetatable** [comment]
i checked on ff developer edition and i can now repro the issue. i hope this commit works

**Comment by apple502j** [comment]
@FunctionalMetatable 
image
image
image

**Comment by FunctionalMetatable** [comment]
menu should properly render hopefully

**Comment by apple502j** [comment]
@FunctionalMetatable 
https://user-images.githubusercontent.com/33279053/129891281-2e2a6ad8-6791-44dc-8cc9-4e689902ee47.mp4

**Comment by apple502j** [comment]
@FunctionalMetatable Newline bug still not fixed for menu?

image

**Comment by FunctionalMetatable** [comment]
> Newline bug still not fixed for menu?

fixed

**Comment by apple502j** [comment]
@FunctionalMetatable Can you support sb menu at signature/post edit/new topic?

**Comment by FunctionalMetatable** [comment]
> @FunctionalMetatable Can you support sb menu at signature/post edit/new topic?

done

File: addons/scratchblocks/userscript.js [hunk 2]
```diff
@@ -30,6 +26,14 @@ async function getLocales(addon) {
     36: "id",
     59: "fa",
   };
+
+  if (location.pathname.split("/")[2] === "settings") {
+    return ["en", locale]
```
Comment (apple502j) [inline addons/scratchblocks/userscript.js]: Vanilla behavior is to use English only, we should do the same

**Comment by apple502j** [comment]
@FunctionalMetatable One more thing and this should be ready to merge, assuming it works :P

File: addons/scratchblocks/userscript.js (hunk 2 above)
Comment (FunctionalMetatable) [reply addons/scratchblocks/userscript.js]: done

File: addons/scratchblocks/userscript.js [hunk 3]
```diff
@@ -76,25 +79,69 @@ export default async function ({ addon, global }) {
...
+    if (document.readyState !== "loading") {
+      resolve();
+    } else {
+      window.addEventListener("DOMContentLoaded", resolve, { once: true });
+    }
+  });
+
+  languages = await getLocales(addon);
+  const blocks = document.querySelectorAll("pre.blocks");
+  if (blocks.length > 0) {
```
Comment (Shluffy) [inline addons/scratchblocks/userscript.js:107]: You could have used `!=` when apple pointed out `!==`

Comment (apple502j) [reply addons/scratchblocks/userscript.js:107]: We banned `!=` operator fyi

Comment (FunctionalMetatable) [reply addons/scratchblocks/userscript.js:107]: it's the same thing

**Pull request merged**